### PR TITLE
WIP: generalize beam related interpreters and packages

### PR DESCRIPTION
**Pull request by LnL7**
###### Motivation for this change

Generalize common configuration for different erlang versions.
- [x] erlang versions
- [ ] erlang feature support
- [x] elixir versions
- [x] hex packages
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by joachifm** [comment]
@ericbmerritt might be interested as well

**Comment by yurrriq** [comment]
Thanks for getting this started. Let us know when it's ready to test and I'll gladly try it on my macOS 10.11 system.

**Comment by LnL7** [comment]
@yurrriq The erlang versions from `beam.interpreter` *should* all work

**Comment by couchemar** [comment]
Thanks you for your work, @LnL7! This is looks very useful

**Comment by binarin** [comment]
Would you also bump R18 to 18.3.4.2?

**Comment by ericbmerritt** [comment]
@LnL7 This is great. I have a branch supporting R19 and I was just lamenting over the copy/paste duplication. I would strongly suggest that you split the refactoring and the version upgrade into separate commits, preferably even separate PRs. Those changes are orthogonal to one another and should be review separately. 

With the refactoring please run `nix-shell -p nox --run "nox-review wip"` thats critical here. There was a recent upgrade to Elixir 1.3 that someone failed to run that for and it broke a fair amount of stuff. I am still in the process of getting all that fixed. I would love if I didn't have to clean up a mess like that again.

**Comment by LnL7** [comment]
@ericbmerritt R19 was recently added to master, I was not planning to change any of the versions here. Since this will be a pretty large change it should probably tested by a few people, maybe we can even build it on hydra.

What was the issue with elixir 1.3?

**Comment by ericbmerritt** [comment]
@LnL7 elixir 1.3 changed bix such that the `--no-deps-check` no longer worked consistently. Even with that elixir would try to go out and check the hex registry. That broke a large number of packages. Fortunately or unfortunately, I don't think there are a lot of elixir users on nix, so the impact has been minimal. In 1.3.2 Jose added the `MIX_NO_DEPS` to address the problem. I should have a PR out with an update to the build infrastructure that supports that flag and a new import of hex-packages.

**Comment by LnL7** [comment]
@ericbmerritt The idea was to version elixir just like we do for erlang.

**Comment by yurrriq** [comment]
I'm finally testing this (on macOS 10.11.6) and can't seem to get 19 to build.

I've got a build from the master branch (I think it was my PR) but now I've discovered some sort of openssl/crypto version mismatch even there.

<details>
<summary>I'm trying 18.3.4.3 now with the following patch.</summary>
```diff
diff --git a/pkgs/development/interpreters/erlang/R18.nix b/pkgs/development/interpreters/erlang/R18.nix
index c917331..364b638 100644
--- a/pkgs/development/interpreters/erlang/R18.nix
+++ b/pkgs/development/interpreters/erlang/R18.nix
@@ -12,8 +12,8 @@ let
   };
 
 in mkDerivation {
-  version = "18.3.4";
-  sha256 = "1f8nhybzsdmjvkmkzpjj3wj9jzx8mihlvi6gfp47fxkalansz39h";
+  version = "18.3.4.3";
+  sha256 = "13iadypni5408mgq4jd0pmgz85asmy4ql1cc8l8hbc0xv7y5vx8z";
 
   patches = [
     rmAndPwdPatch
```
</details>
```fish
nix-env -f /path/to/nixpkg -iA beam.packages.erlangR18.erlang
```

File: pkgs/development/interpreters/erlang/R19.nix [hunk 1]
```diff
@@ -1,100 +1,16 @@
...
-      ++ stdenv.lib.optionals stdenv.isDarwin [ Carbon Cocoa ];
-
-  debugInfo = enableDebugInfo;
+{ mkDerivation, fetchurl }:
 
+let
   envAndCpPatch = fetchurl {
-     url = "https://github.com/binarin/otp/commit/9f9841eb7327c9fe73e84e197fd2965a97b639cf.patch";
-     sha256 = "10h5348p6g279b4q01i5jdqlljww5chcvrx5b4b0dv79pk0p0m9f";
+    url = "https://github.com/binarin/otp/commit/9f9841eb7327c9fe73e84e197fd2965a97b639cf.patch";
```
Comment (binarin) [inline pkgs/development/interpreters/erlang/R19.nix]: Accepted upstream, now you can replace `binarin` with `erlang`.

**Comment by yurrriq** [comment]
ping

**Comment by LnL7** [comment]
This was not not done yet, I mailny opened this pr to get some feedback on the approach I used. I'll see if I can find some time to look at this again. But if somebody else would like to use this as a starting point and finish it feel free to do so!

**Comment by yurrriq** [comment]
I might have a go at this sometime soon, if no one beats me to it.

This seems as good a place as any to ask: why can't I run dialyzer on a project that uses crypto with an erts built from the existing `R19.nix`? I rebuilt with `debugInfo = true` (or whatever it is) and confirmed manually that my `crypto` module indeed has an abstract code chunk, but `dialyzer` claims otherwise... I tried setting `ERL_LIBS` appropriately to no avail...

**Comment by yurrriq** [comment]
I'm working on getting this going with the latest changes from master, while adding a generic LFE builder and updating R18, R19 and Elixir versions, but I'm having some trouble wrt changes to `hex-packages.nix`.

It looks like the changes from this PR to master are basically this at beginning:

```diff
< { pkgs, stdenv, callPackage }:
> { stdenv, pkgs, callPackage, overrides ? (self: super: {}) }:

< self: rec {
> let
>   packages = self: rec {
```

and this at the end:

```diff
< }
>   };
> in stdenv.lib.fix' (stdenv.lib.extends overrides packages)
```

I don't immediately understand what that last change does... but when I try to build, e.g. R19 with my fork, I get the following:

```fish
nix-build \
  -I darwin-config=~/.nixpkgs/darwin-configuration.nix \
  ~/.nix-defexpr/nixpkgs \
  -A beam.packages.erlangR19
```

> error: attempt to call something which is not a function but a set, at ~/.nix-defexpr/nixpkgs/pkgs/development/beam-modules/default.nix:19:8

Any ideas?

/cc @LnL7, @ericbmerritt

**Comment by gleber** [comment]
I am currently reviving this PR as a pure refactoring change.

**Comment by LnL7** [comment]
Closing in favour of https://github.com/NixOS/nixpkgs/pull/26381.

**Pull request closed without merging**